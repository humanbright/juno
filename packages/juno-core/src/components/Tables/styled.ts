import { KeyboardEventHandler } from 'react';

import {
  css,
  fakeBorder,
  focusVisibleShadowStyle,
  focusWithin,
  palette2,
  spacing,
  styled,
} from '../../foundation';
import {
  COLUMN_TEXT_ALIGN,
  ORDER_TABLE_BY,
  TABLE_BORDER_TYPE,
  TABLE_STICKY_TYPE,
  TABLE_TYPE,
} from './types';

interface StyledTableCellProps {
  readonly width?: number | string;
  readonly sortDirection: any;
  readonly hasSortKey?: boolean;
  readonly onClickCapture?: KeyboardEventHandler<HTMLTableHeaderCellElement>;
  readonly onKeyDown?: KeyboardEventHandler<HTMLTableHeaderCellElement>;
}

interface StyledTableHeadCellContentProps {
  readonly textAlign?: COLUMN_TEXT_ALIGN;
}

const TableWrapper = styled('div')`
  height: 100%;
  width: 100%;
  overflow: auto;

  background-color: ${palette2('neutral', 'b01')};
  &.${TABLE_BORDER_TYPE.BORDERED} {
    border-radius: 4px;
    border: 1px solid ${palette2('neutral', 'l02')};
  }
`;

interface StyledTableProps {
  readonly minWidth?: number | string;
}

const StyledTable = styled('table')<StyledTableProps>`
  border-collapse: collapse;
  table-layout: fixed;
  width: 100%;
  display: table;
  /* scroll when viewport is less than 960px including the 332px from left navs */
  min-width: ${spacing(157)};

  @media only screen and (max-width: ${spacing(157)}) {
    table-layout: auto;
  }

  &.${TABLE_STICKY_TYPE.STICKY} th {
    position: sticky;
    top: 0;
  }

  thead {
    border-bottom: 1px solid ${palette2('neutral', 'l02')};
    width: 100%;
    position: sticky;
    top: 0;
    display: table-header-group;
    z-index: 5;
  }

  tbody {
    display: table-row-group;
  }

  tr {
    display: table-row;
    padding-left: ${spacing(4)};

    th,
    td {
      -webkit-box-sizing: border-box;
      -moz-box-sizing: border-box;
      box-sizing: border-box;
      vertical-align: middle;
    }
    td:first-child {
      padding-left: ${spacing(4)};
      margin-left: -${spacing(4)};
    }
    td:not(:first-child) {
      padding-left: ${spacing(4)};
      padding-right: ${spacing(4)};
    }
  }

  &.${TABLE_TYPE.MODAL} tr td {
    overflow: hidden;
    height: 40px;
  }

  &.${TABLE_TYPE.CARD} tr td {
    overflow: hidden;
    height: 52px;
  }

  &.${TABLE_TYPE.FULL} tr td {
    overflow: hidden;
    height: 64px;
  }

  &.${TABLE_TYPE.AUTO} tr td {
    overflow: hidden;
    padding: ${spacing(3, 4)};
    height: auto;
  }
`;

const StyledTableRow = styled('tr')`
  height: 25px;
  display: table-row;
  font-size: 12px;
  text-align: left;
  vertical-align: middle;
`;

const StyledSortIconWrapper = styled.div`
  padding-left: 8px;
  padding-right: 8px;
  vertical-align: middle;
  display: inline-flex;

  & path {
    fill: ${palette2('neutral', 'f04')};
  }
`;

const StyledTableHeadText = styled.div`
  text-overflow: ellipsis;
  overflow: auto;
`;

const StyledTableHeadCell = styled('th')<StyledTableCellProps>`
  color: ${(props) =>
    props.sortDirection === ORDER_TABLE_BY.NONE
      ? palette2('neutral', 'f06')
      : palette2('interactive', 'f01')};
  user-select: none;
  text-align: left;
  background-color: ${(props) =>
    props.sortDirection === ORDER_TABLE_BY.NONE
      ? palette2('neutral', 'b02')
      : palette2('neutral', 'b03')};
  z-index: 100;
  line-height: 32px;
  width: ${(props) => props.width};
  padding-top: 0;
  padding-bottom: 0;
  display: table-cell;
  white-space: nowrap;
  overflow: hidden;

  & > div {
    border: 1px solid inherit;
    padding-left: ${spacing(4)};
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
  }

  ${({ hasSortKey }) =>
    hasSortKey &&
    css`
      ${focusWithin} {
        background-color: ${palette2('neutral', 'b03')};
        ${fakeBorder()};
        cursor: pointer;
        outline: none;
      }
      position: relative;
      ${focusVisibleShadowStyle()};
    `}
`;

// for keep vscode styled-component color
type P = StyledTableHeadCellContentProps;

const StyledTableHeadCellContent = styled.div<P>`
  display: flex;
  align-items: center;

  flex-direction: ${({ textAlign }) =>
    textAlign === COLUMN_TEXT_ALIGN.RIGHT ? 'row-reverse' : 'row'};
  padding-left: ${({ textAlign }) =>
    textAlign === COLUMN_TEXT_ALIGN.RIGHT ? 0 : spacing(4)};
  padding-right: ${({ textAlign }) =>
    textAlign === COLUMN_TEXT_ALIGN.RIGHT ? spacing(4) : 0};
`;

export {
  TableWrapper,
  StyledTable,
  StyledTableRow,
  StyledSortIconWrapper,
  StyledTableHeadCell,
  StyledTableHeadText,
  StyledTableHeadCellContent,
};
