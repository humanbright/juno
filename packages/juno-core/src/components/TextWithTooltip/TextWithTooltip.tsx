import React, { useRef, useState } from 'react';

import {
  getParsePaletteColor,
  palette2,
  RcPaletteProp,
  styled,
  useDeprecatedLog,
} from '../../foundation';

const StyledText = styled.span<{
  textColor: RcPaletteProp;
}>`
  color: ${({ textColor }) =>
    getParsePaletteColor(textColor, palette2('neutral', 'f06'))};
`;

/** @deprecated please don't use that component, just use `RcText` with `title` directly */
const RcTextWithTooltip = (props: {
  children: React.ReactNode;
  tooltip?: string;
  textColor?: RcPaletteProp;
}) => {
  useDeprecatedLog({
    component: 'RcTextWithTooltip',
    message:
      "please don't use that component, just use `RcText` with `title` directly",
  });

  const { tooltip, children, textColor = ['neutral', 'f06'] } = props;
  const ref = useRef<HTMLSpanElement>(null);
  const [showTooltip, setShowTooltip] = useState(false);

  const onMouseOver = () => {
    const span = ref.current;
    if (span && span.parentElement) {
      const { scrollHeight, scrollWidth, clientHeight, clientWidth } =
        span.parentElement;
      setShowTooltip(scrollWidth > clientWidth || scrollHeight > clientHeight);
    }
  };

  const toolTipText = typeof children === 'string' ? children : tooltip || '';

  return (
    <span
      title={showTooltip ? toolTipText : ''}
      onMouseOver={onMouseOver}
      ref={ref}
    >
      <StyledText textColor={textColor}>{children}</StyledText>
    </span>
  );
};

export { RcTextWithTooltip };
