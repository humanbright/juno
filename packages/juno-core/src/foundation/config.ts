/* eslint-disable import/no-mutable-exports */
import { ShowDeprecatedArgs } from './utils';

/* eslint-disable no-console */
const JUNO_WARNING_IGNORE_KEY = '__JUNO_WARNING_IGNORE__';

type WarningFunctionType =
  | ((
      /** warning message */
      message: string,
      /** addition options with that warning */
      options?: Pick<ShowDeprecatedArgs, 'level'>,
    ) => void)
  // * make that type be optional to avoid production not have that method error
  | undefined;

interface RcConfigurationType {
  /** is ignore all warning show from juno */
  WARNING_IGNORE: boolean;
  WARNING_FUNCTION: WarningFunctionType;
}

// * use that way to avoid that code be exist in production code base
/**
 * config global setting
 *
 * ```ts
 * configure({
 *   warning: false
 * })
 * ```
 * ## WARNING related method and  only work in `development mode`
 *
 * `WARNING_IGNORE`:  only work in `development mode`
 * `WARNING_FUNCTION` only work in `development mode`
 */
let rcConfiguration = {} as RcConfigurationType;

if (process.env.NODE_ENV !== 'production') {
  class RcConfig implements RcConfigurationType {
    private _WARNING_IGNORE: boolean = false;

    private _WARNING_FUNCTION: WarningFunctionType = (
      message: string,
      options,
    ) => {
      console[options?.level || 'warn'](message);
    };

    constructor() {
      if (typeof window !== 'undefined') {
        // SSR support
        // * get value once when init
        this._WARNING_IGNORE =
          window?.localStorage?.getItem(JUNO_WARNING_IGNORE_KEY) === 'true';

        // * also provide user to set variable from window
        (window as any).__JUNO__ = this;
      }
    }

    set WARNING_IGNORE(value: boolean) {
      if (value) {
        console.log('---------------- [Juno] ----------------');
        console.log(
          `Don't recommended you close it, we should avoid use any deprecated props or methods.`,
        );
        console.log('---------------- [Juno] ----------------');
      }

      this._WARNING_IGNORE = !!value;

      window?.localStorage?.setItem(
        JUNO_WARNING_IGNORE_KEY,
        `${this._WARNING_IGNORE}`,
      );
    }

    get WARNING_IGNORE() {
      return this._WARNING_IGNORE;
    }

    set WARNING_FUNCTION(func: WarningFunctionType) {
      if (typeof func === 'function') {
        this._WARNING_FUNCTION = func;
      } else {
        console.warn('[Juno] config warningFunction is not a function.');
      }
    }

    get WARNING_FUNCTION() {
      return this._WARNING_FUNCTION;
    }
  }

  rcConfiguration = new RcConfig();
}

type RcConfigure = {
  /** is show warning message in development */
  warning?: boolean;
  /** custom warning callback */
  warningFunction?: WarningFunctionType;
};

/**
 * set config with Juno
 *
 * ## WARNING method only work in `development mode`
 */
export const configure = (options: RcConfigure) => {
  if (process.env.NODE_ENV !== 'production') {
    const { warning, warningFunction } = options;
    if (warning !== undefined) {
      rcConfiguration.WARNING_IGNORE = !warning;
    }
    if (warningFunction !== undefined) {
      rcConfiguration.WARNING_FUNCTION = warningFunction;
    }
  }
};

export { rcConfiguration };
export type { WarningFunctionType };
