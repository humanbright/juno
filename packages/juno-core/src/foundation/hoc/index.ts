export * from './withDelay';
export * from './withResponsive';
