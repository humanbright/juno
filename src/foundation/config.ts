import { ShowDeprecatedArgs } from './utils';

/* eslint-disable no-console */
const JUNO_WARNING_IGNORE_KEY = '__JUNO_WARNING_IGNORE__';

type WarningFunctionType = (
  /** warning message */
  message: string,
  /** addition options with that warning */
  options?: Pick<ShowDeprecatedArgs, 'level'>,
) => void;

class RcConfig {
  private _WARNING_IGNORE: boolean = false;

  private _WARNING_FUNCTION: WarningFunctionType = (
    message: string,
    options,
  ) => {
    console[options?.level || 'warn'](message);
  };

  constructor() {
    if (typeof window !== 'undefined') {
      // SSR support
      // * get value once when init
      this._WARNING_IGNORE =
        window?.localStorage?.getItem(JUNO_WARNING_IGNORE_KEY) === 'true';

      // * also provide user to set variable from window
      (window as any).__JUNO__ = this;
    }
  }

  set WARNING_IGNORE(value: boolean) {
    if (value) {
      console.log('---------------- [Juno] ----------------');
      console.log(
        `Don't recommended you close it, we should avoid use any deprecated props or methods.`,
      );
      console.log('---------------- [Juno] ----------------');
    }

    this._WARNING_IGNORE = !!value;

    window?.localStorage?.setItem(
      JUNO_WARNING_IGNORE_KEY,
      `${this._WARNING_IGNORE}`,
    );
  }

  get WARNING_IGNORE() {
    return this._WARNING_IGNORE;
  }

  set WARNING_FUNCTION(func: WarningFunctionType) {
    if (typeof func === 'function') {
      this._WARNING_FUNCTION = func;
    } else {
      console.warn('[Juno] config warningFunction is not a function.');
    }
  }

  get WARNING_FUNCTION() {
    return this._WARNING_FUNCTION;
  }
}

/** config instance */
const rcConfiguration = new RcConfig();

type RcConfigure = {
  /** is show warning message in development */
  warning?: boolean;
  /** custom warning callback */
  warningFunction?: WarningFunctionType;
};

/**
 * set config with Juno
 */
export const configure = ({ warning, warningFunction }: RcConfigure) => {
  if (warning !== undefined) {
    rcConfiguration.WARNING_IGNORE = !warning;
  }
  if (warningFunction !== undefined) {
    rcConfiguration.WARNING_FUNCTION = warningFunction;
  }
};

export { rcConfiguration };
export type { WarningFunctionType };
