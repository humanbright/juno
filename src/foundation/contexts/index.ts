export * from './PortalWindowContext';
export * from './ResponsiveContext';
