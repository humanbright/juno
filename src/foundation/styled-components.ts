import { useTheme } from '@material-ui/core/styles';
/* eslint-disable import/no-duplicates */
import * as styledComponents from 'styled-components';

import {
  ThemedStyledComponentsModule,
  ThemeProps as StyledThemeProps,
  InterpolationValue,
  Keyframes,
  Interpolation,
  FlattenInterpolation,
  StyleSheetManager,
  ThemedStyledProps,
} from 'styled-components';

import { RcTheme } from './theme/theme.type';

type ThemeProps = StyledThemeProps<RcTheme>;

const {
  default: styled,
  css,
  keyframes,
  createGlobalStyle,
  withTheme,
  ThemeProvider,
  ThemeConsumer,
} = (styledComponents as any) as ThemedStyledComponentsModule<RcTheme>;

type Dependencies = {
  dependencies?: (React.ComponentType | ((props: any) => JSX.Element))[];
};

const RcUseTheme = <T = RcTheme>() => useTheme<T>();

export {
  css,
  createGlobalStyle,
  keyframes,
  ThemeProvider,
  withTheme,
  ThemeConsumer,
  ThemeProps,
  Dependencies,
  InterpolationValue,
  Keyframes,
  FlattenInterpolation,
  Interpolation,
  RcUseTheme as useTheme,
  ThemedStyledProps,
  StyleSheetManager,
};

export default styled;
