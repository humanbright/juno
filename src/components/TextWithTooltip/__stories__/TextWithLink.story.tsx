import { number, text } from '@storybook/addon-knobs';
import React from 'react';

import { ellipsis } from '../../../foundation';
import styled from '../../../foundation/styled-components';
import { RcTextWithTooltip } from '../TextWithTooltip';

export default {
  title: 'ðŸ–¤ Deprecated Components/Text/TextWithTooltip',
};

const Parent = styled.div`
  ${ellipsis()};
`;

export const TextWithTooltip = () => {
  const width = number('width', 400);
  const content = text(
    'text',
    'If the text is truncated, will show tooltip when hovered. hahahahahahah',
  );
  return (
    <Parent style={{ width: `${width}px` }}>
      <RcTextWithTooltip>{content}</RcTextWithTooltip>
    </Parent>
  );
};

TextWithTooltip.story = {
  name: 'TextWithTooltip',
};
