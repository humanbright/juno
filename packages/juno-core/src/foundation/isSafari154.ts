import { createGlobalStyle } from 'styled-components';

import { RcClasses, RcClassesProps } from './utils/classes';
import { combineProps } from './utils/combineProps';

/**
 * Conditionally apply a workaround for the CSS transition bug in Safari 15.4.
 * Remove this workaround once the Safari bug is fixed.
 */
export const isSafari154 =
  typeof navigator !== 'undefined' &&
  (/^((?!chrome|android).)*safari/i.test(navigator.userAgent) ||
    /(iPhone|iPad)/.test(navigator.userAgent)) &&
  /15\.[4-9]/i.test(navigator.userAgent);

const FixSafari154Classes = RcClasses<RcClassesProps<'paper'>>(
  ['paper'],
  'RcFixSafari154',
);

/**
 * Conditionally apply a workaround for the CSS transition bug in Safari 15.4.
 * Remove this workaround once the Safari bug is fixed.
 */
export const getSafari154Theme = (theme: any, enable?: boolean) => {
  if (enable ?? isSafari154) {
    return combineProps(
      {
        props: {
          // * set all grow related to only have one transition to make animation not shake after safari 15.4
          MuiMenu: {
            PopoverClasses: { paper: FixSafari154Classes.paper },
          },
          RcVirtualizedMenu: {
            PopoverClasses: { paper: FixSafari154Classes.paper },
          },
          MuiPopover: {
            PaperProps: { className: FixSafari154Classes.paper },
          },
          // mui tooltip prop will cover all existing props
          RcTooltip: {
            classes: {
              tooltip: FixSafari154Classes.paper,
            },
          },
          // mui grow not accept global prop
          RcGrow: {
            className: FixSafari154Classes.paper,
          },
        },
      },
      theme,
    );
  }

  return theme;
};

export const fixSafariTransitionStyle =
  'all 159ms cubic-bezier(0.4, 0, 0.2, 1) 0ms';
/**
 * A global style that can be used to apply a workaround for the CSS transition bug in Safari 15.4.
 */
export const GlobalFixSafariStyle = createGlobalStyle`
  .${FixSafari154Classes.paper} {
    transition: ${fixSafariTransitionStyle} !important;
  }
`;
