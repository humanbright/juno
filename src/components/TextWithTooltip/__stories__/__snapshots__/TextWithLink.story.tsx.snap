// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ðŸ–¤ Deprecated Components/Text/TextWithTooltip TextWithTooltip 1`] = `
.c1 {
  color: #121212;
}

.c0 {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

<div
  className="c0"
  style={
    Object {
      "width": "400px",
    }
  }
>
  <span
    onMouseOver={[Function]}
    title=""
  >
    <span
      className="c1"
    >
      If the text is truncated, will show tooltip when hovered. hahahahahahah
    </span>
  </span>
</div>
`;
